thalo-eventstoredb/src/event_store.rs

```diff
@@ -0,0 +1,340 @@
+        }
+
+        let sequence = self.load_aggregate_sequence::<A>(id).await.unwrap_or(None);
+
+        let revision = {
+            match sequence {
+                Some(n) => ExpectedRevision::Exact(n as u64),
+                None => ExpectedRevision::NoStream,
+            }
+        };
+
+        let mut payload: Vec<EventData> = vec![];
+
+        for event in events.iter() {
+            let event_data_payload = ESDBEventPayload {
+                created_at: Utc::now(),
+                aggregate_type: <A as TypeId>::type_id().to_string(),
+                aggregate_id: id.to_string(),
+                event_data: serde_json::to_value(event).map_err(Error::SerializeEvent)?,
+            };
+
+            let event_data = EventData::json(event.event_type(), &event_data_payload)
+                .map_err(Error::SerializeEventDataPayload)?
+                .id(Uuid::new_v4());
+
+            payload.push(event_data);
+        }
+
+        let options = AppendToStreamOptions::default().expected_revision(revision);
+        let res = self
+            .client
+            .append_to_stream(&self.stream_id::<A>(Some(id)), &options, payload)
+            .map_err(|err| Error::WriteStreamError(sequence.unwrap_or(0) as usize, err))
+            .await?;
+
+        let end_position = res.next_expected_version as usize;
+        let start_position = end_position - (events.len() - 1);
+
+        Ok((start_position..end_position + 1).collect::<Vec<usize>>())
+    }
+}
+
+impl Debug for ESDBEventStore {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("ESDBEventStore")
+            .field("client", &"eventstore::Client")
+            .finish()
+    }
+}
+
+#[cfg(feature = "debug")]
+impl ESDBEventStore {
+    pub async fn print<A>(&self)
+    where
+        A: Aggregate,
+        <A as Aggregate>::Event: serde::Serialize,
+    {
+        let stream_res = self.client.read_all(&Default::default()).await;
+
+        let mut events: Vec<(Uuid, u64, ESDBEventPayload)> = vec![];
+        if let Ok(mut stream) = stream_res {
+            while let Some(event) = stream.next().await.unwrap() {
+                let event_data = event.get_original_event();
+
+                if event_data.event_type.starts_with('$') {
+                    continue;
+                }
+
+                let is_aggregate_event = event_data
+                    .stream_id
+                    .starts_with(self.stream_id::<A>(None).as_str());
+
+                if !is_aggregate_event {
+                    continue;
+                }
+
+                // TODO: Make ESDBEventPayload handle all this
+                // and implement fns for RecordedEvent -> ESDBEventPayload, ESDBEventPayload -> AggregateEnvelope, ESDBPayload -> eventstore::EventData
+                // make base struct handle id and revision seperately - update event trait for different id types (str, u64, uuid) etc
+                events.push((
```

Comment:
Nice to see you added the table debug info :)
There are a few TODO comments, are they still planned to be done before the PR is merged?

Reply:
I'll evaluate my TODOs, We should probably add some sort of pagination to this fn if its going to be used anywhere outside of the /examples. Any real world example is going to end up being an impractical size with much activity